### PR TITLE
Reorganize modules and some of the code within them

### LINKED ISSUE #306: Reorganize modules
Things have grown rather organically. If we're doing a big breaking 1.0 release, are there any reshuffling of imports we want to do?

### PR DESCRIPTION
**Pull request by spladug**
This is a biggggg reshuffle and will require a lot of updates in
applications which we will do with automatic tooling.

Some of these are simple renames (context -> clients, integrations ->
frameworks, / -> lib), but there are also some subtler shifts like
moving the core stuff up to root `baseplate` and pulling edge context
out into its own lib.

The overall goal is to make the layout of the library much more obvious
to newcomers and to help improve the cleanliness of imports in
applications.

Here's the full list of remappings from the tool I used to rewrite all references in the code/docs and will make available to service owners.

```
baseplate._utils                            → baseplate.lib                                          
baseplate._utils.Batch                      → baseplate.sidecars.Batch                               
baseplate._utils.BatchFull                  → baseplate.sidecars.BatchFull                           
baseplate._utils.RawJSONBatch               → baseplate.sidecars.RawJSONBatch                        
baseplate._utils.SerializedBatch            → baseplate.sidecars.SerializedBatch                     
baseplate._utils.TimeLimitedBatch           → baseplate.sidecars.TimeLimitedBatch                    
baseplate.config                            → baseplate.lib.config                                   
baseplate.context                           → baseplate.clients                                      
baseplate.core                              → baseplate                                              
baseplate.core.AuthenticationToken          → baseplate.lib.edge_context.AuthenticationToken         
baseplate.core.AuthenticationTokenValidator → baseplate.lib.edge_context.AuthenticationTokenValidator
baseplate.core.EdgeRequestContext           → baseplate.lib.edge_context.EdgeRequestContext          
baseplate.core.EdgeRequestContextFactory    → baseplate.lib.edge_context.EdgeRequestContextFactory   
baseplate.core.InvalidAuthenticationToken   → baseplate.lib.edge_context.InvalidAuthenticationToken  
baseplate.core.NoAuthenticationError        → baseplate.lib.edge_context.NoAuthenticationError       
baseplate.core.OAuthClient                  → baseplate.lib.edge_context.OAuthClient                 
baseplate.core.Service                      → baseplate.lib.edge_context.Service                     
baseplate.core.Session                      → baseplate.lib.edge_context.Session                     
baseplate.core.User                         → baseplate.lib.edge_context.User                        
baseplate.core.ValidatedAuthenticationToken → baseplate.lib.edge_context.ValidatedAuthenticationToken
baseplate.crypto                            → baseplate.lib.crypto                                   
baseplate.crypto.constant_time_compare      → hmac.compare_digest                                    
baseplate.datetime                          → baseplate.lib.datetime                                 
baseplate.diagnostics                       → baseplate.observers                                    
baseplate.diagnostics.tracing.publisher     → baseplate.sidecars.trace_publisher                     
baseplate.error_reporter_from_config        → baseplate.observers.sentry.error_reporter_from_config  
baseplate.events                            → baseplate.lib.events                                   
baseplate.events.publisher                  → baseplate.sidecars.event_publisher                     
baseplate.events.queue                      → baseplate.lib.events                                   
baseplate.experiments                       → baseplate.lib.experiments                              
baseplate.file_watcher                      → baseplate.lib.file_watcher                             
baseplate.integration                       → baseplate.frameworks                                   
baseplate.live_data                         → baseplate.lib.live_data                                
baseplate.live_data.watcher                 → baseplate.sidecars.live_data_watcher                   
baseplate.message_queue                     → baseplate.lib.message_queue                            
baseplate.metrics                           → baseplate.lib.metrics                                  
baseplate.metrics_client_from_config        → baseplate.lib.metrics.metrics_client_from_config       
baseplate.queue_consumer                    → baseplate.frameworks.queue_consumer                    
baseplate.random                            → baseplate.lib.random                                   
baseplate.ratelimit                         → baseplate.lib.ratelimit                                
baseplate.requests                          → baseplate.lib._requests                                
baseplate.retry                             → baseplate.lib.retry                                    
baseplate.secrets                           → baseplate.lib.secrets                                  
baseplate.secrets.fetcher                   → baseplate.sidecars.secrets_fetcher                     
baseplate.secrets.store                     → baseplate.lib.secrets                                  
baseplate.service_discovery                 → baseplate.lib.service_discovery                        
baseplate.thrift_pool                       → baseplate.lib.thrift_pool                              
baseplate.tracing_client_from_config        → baseplate.observers.tracing.tracing_client_from_config 
```

Fixes #306

### CODE REVIEW

**Comment by spladug** [comment]
:eyeglasses: @pacejackson  @bradengroom

File: baseplate/__init__.py [hunk 1]
```diff
@@ -1,215 +1,624 @@
...
+logger = logging.getLogger(__name__)
 
 
-def metrics_client_from_config(raw_config: config.RawConfig) -> metrics.Client:
-    """Configure and return a metrics client.
+class BaseplateObserver:
+    """Interface for an observer that watches Baseplate."""
 
-    This expects two configuration options:
+    def on_server_span_created(self, context: Any, server_span: "Span") -> None:
```
Comment (bradengroom) [inline baseplate/__init__.py]: Shouldn't we be using the `RequestContext` type here instead of `Any`?

Comment (spladug) [reply baseplate/__init__.py]: Probably, now that it exists. I'm going to do another pass of typing after this is merged and will make sure to fix that up. GP!

**Comment by spladug** [comment]
> Should we do a pass and clean up the imports? They’re not really ordered nicely anymore.

Yup. Coming next: `isort` or similar applied to the whole lib.

File: baseplate/sidecars/__init__.py [hunk 2]
```diff
@@ -1,44 +1,5 @@
...
-
-    def __get__(self, instance, owner):
-        if instance is None:
-            return self
-        ret = self.wrapped(instance)
-        setattr(instance, self.wrapped.__name__, ret)
-        return ret
 
 
 SerializedBatch = collections.namedtuple("SerializedBatch", "count bytes")
```
[the remark this answers is not in the document]
Comment (spladug) [reply baseplate/sidecars/__init__.py]: Where would be better? Like a `baseplate.sidecars._utils`? :)

[the remark this answers is not in the document]
Comment (pacejackson) [reply baseplate/sidecars/__init__.py]: That works, or maybe `baseplate.sidecars.batch`. That said, it’s a fairly nitpicky distinction and I don’t hate having it here.

**Comment by spladug** [comment]
I'll come back to the tests. They should definitely be cleaner.

**Pull request merged**